src/main/java/com/uid2/admin/vertx/service/SiteService.java

```diff
@@ -196,4 +208,75 @@ private void handleSiteEnable(RoutingContext rc) {
             rc.fail(500, e);
         }
     }
+
+    private void handleSiteDomains(RoutingContext rc) {
+        try {
+            // refresh manually
+            siteProvider.loadContent();
+
+            final Site existingSite = RequestUtil.getSite(rc, "id", siteProvider);
+            if (existingSite == null) {
+                return;
+            }
+
+            JsonObject body = rc.body().asJsonObject();
+            JsonArray domainNamesJa = body.getJsonArray("domain_names");
+            if(domainNamesJa == null) {
+                ResponseUtil.error(rc, 400, "required parameters: domain_names");
+                return;
+            }
+            List<String> domainNames = domainNamesJa.stream().map(String::valueOf).collect(Collectors.toList());
+
+            List<String> normalizedDomainNames = new ArrayList<>();
+            for(String domain : domainNames) {
+                try {
+                    String tld = getTopLevelDomainName(domain);
+                    normalizedDomainNames.add(tld);
+                } catch (Exception e) {
+                    ResponseUtil.error(rc, 400, "invalid domain name: " + domain);
+                    return;
+                }
+            }
+
+            boolean containsDuplicates = normalizedDomainNames.stream().distinct().count() < normalizedDomainNames.size();
+            if (containsDuplicates) {
+                ResponseUtil.error(rc, 400, "duplicate domain_names not permitted");
+                return;
+            }
+
+            existingSite.setDomainNames(new HashSet<>(normalizedDomainNames));
+
+            final List<Site> sites = this.siteProvider.getAllSites()
+                    .stream().sorted(Comparator.comparingInt(Site::getId))
+                    .collect(Collectors.toList());
+
+            storeWriter.upload(sites, null);
+
+            rc.response().end(jsonWriter.writeValueAsString(existingSite));
+        } catch (Exception e) {
+            ResponseUtil.errorInternal(rc, "set site domain_names failed", e);
+        }
+    }
+
+    public static String getTopLevelDomainName(String origin) throws MalformedURLException {
+        String host;
+        try {
+            URL url = new URL(origin);
+            host = url.getHost();
+        } catch (Exception e) {
+            host = origin;
+        }
+        //InternetDomainName will normalise the domain name to lower case already
+        InternetDomainName name = InternetDomainName.from(host);
```

Comment:
i'm surprised this works without changing pom.xml to explicitly add a google guava dependency?

Reply:
hmm it works fine for me without it but I will add it in if that's just me.

Reply:
I accidentally put the google dependency in the ui branch. I have added it to this branch now and also merged master into my shared branch. No build issues for me